### PR TITLE
Migration 

### PR DESCRIPTION
**Pull request by macpie**
Fix https://github.com/helium/router/issues/869

### CODE REVIEW

**Comment by macpie** [comment]
I will match payload against API better once they are most set in stone

File: src/cli/router_cli_migration.erl [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                EUIS ++ [RouteSpacer]
+        end,
+        [io_lib:format("Routes~n", [])],
+        maps:get(routes, Map)
+    ),
+    c_list([Disclamer, OUI, Owner, Payer] ++ Routes).
+
+-spec euis() -> list(map()).
+euis() ->
+    Devices = router_device_cache:get(),
```
Comment (michaeldjeffrey) [inline src/cli/router_cli_migration.erl:214]: Should we check Console so we don't have stale devices.

File: src/cli/router_cli_migration.erl [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+                    Payer = blockchain_swarm:pubkey_bin(),
+                    Map = #{
+                        oui => OUI,
+                        payer => erlang:list_to_binary(libp2p_crypto:bin_to_b58(Owner)),
+                        owner => erlang:list_to_binary(libp2p_crypto:bin_to_b58(Payer)),
+                        routes => [
+                            #{
+                                devaddr_ranges => DevAddrRanges,
+                                euis => EUIs,
+                                net_id => erlang:list_to_binary(io_lib:format("~.16B", [IntNetID])),
```
Comment (michaeldjeffrey) [inline src/cli/router_cli_migration.erl:255]:
```suggestion
                                net_id => erlang:list_to_binary(io_lib:format("~6.16.0B", [IntNetID])),
```

6 wide, fill with `0`.
But I they should all be the same, yeah?

File: config/sys.config.src [hunk 3]
```diff
@@ -45,7 +45,10 @@
             {downlink_endpoint, <<"${ROUTER_CONSOLE_DOWNLINK_ENDPOINT}">>},
             {secret, <<"${ROUTER_CONSOLE_SECRET}">>}
         ]},
-        {router_xor_filter_worker, "${ROUTER_XOR_FILTER_WORKER}"},
+        {ics, #{
+            host => "${ROUTER_ICS_HOST}"
+            port => ${ROUTER_ICS_PORT:-8080}
```
Comment (michaeldjeffrey) [inline config/sys.config.src:50]: Does this syntax work for you? I can never get rebar3 to not complain about it when I launch a shell.

File: src/router_sup.erl [hunk 4]
```diff
@@ -137,7 +142,7 @@ init([]) ->
             ?SUP(router_console_sup, []),
             ?SUP(router_decoder_sup, []),
             ?WORKER(router_device_devaddr, [#{}]),
-            ?WORKER(router_xor_filter_worker, [#{}])
```
Comment (michaeldjeffrey) [inline src/router_sup.erl:140]: yltsi 🕺

File: src/cli/router_cli_migration.erl [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+send_euis_to_config_service(["migration", "euis"], [], Flags) ->
+    Options = maps:from_list(Flags),
+
+    PubKeyBin = blockchain_swarm:pubkey_bin(),
+    {ok, _, SigFun, _} = blockchain_swarm:keys(),
+
+    RouteEuisReq = #{
+        id => maps:get(route_id, Options, ""),
+        action => add,
```
Comment (michaeldjeffrey) [inline src/cli/router_cli_migration.erl:158]:
```suggestion
        action => add_euis,
```

File: src/cli/router_cli_migration.erl [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+                            }
+                        ]
+                    },
+                    {ok, Map}
+            end
+    end.
+
+-spec euis() -> list(map()).
+euis() ->
+    {ok, Devices} = get_devices(),
```
Comment (michaeldjeffrey) [inline src/cli/router_cli_migration.erl:275]:
```suggestion
    {ok, Devices} = router_console_api:get_json_devices(),
```

And all the api stuff below can be dropped?

File: src/router_ics_worker.erl [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    pubkey_bin :: libp2p_crypto:pubkey_bin(),
+    sig_fun :: function(),
+    host :: string(),
+    port :: non_neg_integer(),
+    conn :: undefined | grpc_client:connection(),
+    conn_backoff :: backoff:backoff(),
+    route_id :: undefined | string()
+}).
+
+-type state() :: state().
```
Comment (michaeldjeffrey) [inline src/router_ics_worker.erl:48]:
```suggestion
-type state() :: #state{}.
```

File: src/router_ics_worker.erl [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    case euis_req(Conn, PubKeyBin, SigFun, RouteID, add_euis, ToAdd) of
+        true -> lager:info("added ~p", [ToAdd]);
+        false -> lager:warning("failed to add ~p", [ToAdd])
+    end,
+    {noreply, State};
+handle_cast(
+    {remove, DeviceIDs},
+    #state{pubkey_bin = PubKeyBin, sig_fun = SigFun, conn = Conn, route_id = RouteID} = State
+) ->
+    Euis = fetch_device_euis(cache, DeviceIDs),
```
Comment (michaeldjeffrey) [inline src/router_ics_worker.erl:131]: Should we also remove these devices from the cache?

File: src/router_sup.erl [hunk 9]
```diff
@@ -137,7 +142,8 @@ init([]) ->
...
+            ?WORKER(router_xor_filter_worker, [#{}]),
+            ?WORKER(router_ics_worker, [ICSOpts])
```
Comment (michaeldjeffrey) [inline src/router_sup.erl:146]: Should this be started before the `console_sup` so we don't miss any messages?

Comment (michaeldjeffrey) [reply src/router_sup.erl:146]: Hmm, well it needs the console worker to make the refetch request when it starts up. Nevermind.

File: src/router_ics_worker.erl (hunk 8 above)
Comment (macpie) [reply src/router_ics_worker.erl:131]: yep

**Review by michaeldjeffrey — APPROVED** [review]

**Pull request merged**